File: zig/3-pid-bm/main/Registers.zig
const std = @import("std");

const Self = @This();

input: Input = .{},
holding: Holding = .{},

pub const Input = packed struct {
    frequency: FloatCDAB = FloatCDAB.zero,
    control_signal: FloatCDAB = FloatCDAB.zero,
};

pub const Holding = packed struct {
    target_frequency: FloatCDAB = FloatCDAB.zero,
    proportional_factor: FloatCDAB = FloatCDAB.zero,
    integration_time: FloatCDAB = FloatCDAB.inf,
    differentiation_time: FloatCDAB = FloatCDAB.zero,
};

pub const FloatCDAB = packed struct {
    const zero = FloatCDAB.fromF32(0.0);
    const inf = FloatCDAB.fromF32(std.math.inf(f32));

    raw: packed struct { c: u8, d: u8, a: u8, b: u8 },

    pub fn fromF32(value: f32) FloatCDAB {
        const bytes: [4]u8 = @bitCast(value);
        return .{ .raw = .{
            .c = bytes[2],
            .d = bytes[3],
            .a = bytes[0],
            .b = bytes[1],
        } };
    }

    pub fn toF32(self: FloatCDAB) f32 {
        const bytes = [4]u8{ self.raw.a, self.raw.b, self.raw.c, self.raw.d };
        return @bitCast(bytes);
    }
};

File: rust/3-pid/src/registers.rs
use std::{
    mem::variant_count,
    ops::{Bound, IntoBounds},
};

use num_derive::{FromPrimitive, ToPrimitive};
use num_traits::{FromPrimitive, ToPrimitive};

#[derive(PartialOrd, Ord, PartialEq, Eq, Clone, Copy, FromPrimitive, ToPrimitive)]
pub enum InputRegister {
    Frequency,
    ControlSignal,
}

#[derive(PartialOrd, Ord, PartialEq, Eq, Clone, Copy, FromPrimitive, ToPrimitive)]
pub enum HoldingRegister {
    TargetFrequency,
    ProportionalFactor,
    IntegrationTime,
    DifferentiationTime,
}

pub struct Registers {
    input: [f32; variant_count::<InputRegister>()],
    holding: [f32; variant_count::<HoldingRegister>()],
}

impl Registers {
    pub fn new() -> Self {
        Self {
            input: Default::default(),
            holding: [0., 0., f32::INFINITY, 0.],
        }
    }

    pub fn read_input_registers(&self, range: impl IntoBounds<InputRegister>) -> &[f32] {
        &self.input[into_usize_bounds(range)]
    }

    pub fn write_input_registers(
        &mut self,
        range: impl IntoBounds<InputRegister>,
        values: impl IntoIterator<Item = f32>,
    ) {
        self.input[into_usize_bounds(range)]
            .iter_mut()
            .zip(values)
            .for_each(|(reg, value)| *reg = value);
    }

    pub fn read_holding_registers(&self, range: impl IntoBounds<HoldingRegister>) -> &[f32] {
        &self.holding[into_usize_bounds(range)]
    }

    pub fn write_holding_registers(
        &mut self,
        range: impl IntoBounds<HoldingRegister>,
        values: impl IntoIterator<Item = f32>,
    ) {
        self.holding[into_usize_bounds(range)]
            .iter_mut()
            .zip(values)
            .for_each(|(reg, value)| *reg = value);
    }
}

fn into_usize_bounds<T>(range: impl IntoBounds<T>) -> (Bound<usize>, Bound<usize>)
where
    T: ToPrimitive,
{
    let (start, end) = range.into_bounds();
    let start = start.map(|x| x.to_usize().unwrap());
    let end = end.map(|x| x.to_usize().unwrap());
    (start, end)
}

pub fn range_from_addr_count<T>(addr: u16, count: u16) -> Option<impl IntoBounds<T>>
where
    T: FromPrimitive,
{
    let start = T::from_u16(addr)?;
    let end = T::from_u16(addr + count - 1)?;
    Some(start..=end)
}
